### DISCUSSION TITLE
Dynamically set an env variable on deployment

### DISCUSSION THREAD

**Discussion by fabien-huet** [vercel/community#722] [Help] [answered]
To call a third party provider, I need to generate an API token and that token is restricted to allow CORS only to a list of given domains.

It's medium lived (a few months) so the best way is to generate it every time I start an environment.

Locally, a script generates the token with `http://localhost:3000`, then update a local .env, then starts Next.

But when deploying, what's the equivalent ? I can generate the token with `process.env.VERCEL_URL` before building. It works. But then, how to inject it somewhere ?

I tried `vercel env add NEXT_PUBLIC_MY_TOKEN=${token}`. But I get `stderr: /bin/sh: vercel: command not found`. I am going in the right direction ?

Would it be possible to just `fs.writeFile` in the codebase to inject a `secrets.json` file ? Locally, that would work, but on vercel ... ?

**Comment by amyegan** [comment, marked as answer]
If you already [installed Vercel CLI](https://vercel.com/docs/cli#introduction/installing-the-cli) and [linked your Project](https://vercel.com/docs/cli#introduction/project-linking), you should be able to use the `env` command to add the token. Usage is a little different from what you tried though. If you run just `vercel env add`, it will ask you a series of questions to add a new variable. You can skip the questions by using something like `vercel env add NEXT_PUBLIC_MY_TOKEN development < my-token.txt` to set the name, environment, and token value from file all at once.
